### PR TITLE
Ppl count approximate support

### PR DESCRIPTION
**Pull request by YANG-DB**
### Description
support approximation operations for 
 - count distinct
 - top
 - rare
 
### Related Issues
#882 

### related context
- https://spark.apache.org/docs/3.5.2/sql-ref-functions-builtin.html

### Check List
- [x] Updated documentation (docs/ppl-lang/README.md)
- [x] Implemented unit tests
- [X] Implemented tests for combination with other commands
- [X] New added source code should include a copyright header
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ppl-spark-integration/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java [hunk 1]
```diff
@@ -185,6 +185,7 @@ public enum BuiltinFunctionName {
   NESTED(FunctionName.of("nested")),
   PERCENTILE(FunctionName.of("percentile")),
   PERCENTILE_APPROX(FunctionName.of("percentile_approx")),
+  APPROX_COUNT_DISTINCT(FunctionName.of("approx_count_distinct")),
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply ppl-spark-integration/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java:188]: Oh, I see https://github.com/opensearch-project/opensearch-spark/pull/884/files#diff-66576caee2b95944ec89dac941862c61a3e415f999a44ddcf03684f3d7c61115R60, might not need to add mapping item above, but no harmful.

**Review by LantaoJin — APPROVED** [review]
All my comments are minor change requests. LGTM basically.

File: docs/ppl-lang/ppl-top-command.md [hunk 2]
```diff
@@ -19,6 +20,7 @@ The example finds most common gender of all the accounts.
 PPL query:
 
     os> source=accounts | top gender;
+    os> source=accounts_approx | top gender;
```
[the remark this answers is not in the document]
Comment (YANG-DB) [reply docs/ppl-lang/ppl-top-command.md:23]: typing error - thanks !

File: ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4 [hunk 3]
```diff
@@ -400,7 +402,7 @@ statsAggTerm
 statsFunction
    : statsFunctionName LT_PRTHS valueExpression RT_PRTHS                                                                            # statsFunctionCall
    | COUNT LT_PRTHS RT_PRTHS                                                                                                        # countAllFunctionCall
-   | (DISTINCT_COUNT | DC) LT_PRTHS valueExpression RT_PRTHS                                                                        # distinctCountFunctionCall
+   | (DISTINCT_COUNT | DC | DISTINCT_COUNT_APPROX) LT_PRTHS valueExpression RT_PRTHS                                                                        # distinctCountFunctionCall
```
[the remark this answers is not in the document]
Comment (YANG-DB) [reply ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4:405]: yes - we should add this to the github issue check - list ;-)

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystPlanContext.java [hunk 4]
```diff
@@ -188,6 +189,13 @@ public LogicalPlan reduce(BiFunction<LogicalPlan, LogicalPlan, LogicalPlan> tran
         }).orElse(getPlan()));
     }
 
+    /**
+     * update context using the given action and node 
+     */
+    public CatalystPlanContext update(UnaryOperator<CatalystPlanContext> action) {
```
[the remark this answers is not in the document]
Comment (YANG-DB) [reply ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystPlanContext.java:195]: I was thinking of generalize any access to the context under a functional call - look like its too much probably...

**Pull request merged**